version: '3.4'

# Common Django template for Geonode, Celery and Celerycam services below
x-common-django:
  &default-common-django
  image: kartoza/igrac_ggis:${VERSION}
  build:
    context: ../
    dockerfile: deployment/docker/Dockerfile
    target: prod
  env_file:
    - .env
  volumes:
#    - ../geonode:/geonode
    - ../django_project:/home/web/django_project
    - ./volumes/statics:/home/web/statics:rw
  restart: on-failure
  depends_on:
    - db

services:
  # Vanilla RabbitMQ service. This is needed by celery
  rabbitmq:
    image: rabbitmq:3.7-alpine
    volumes:
      - rabbitmq:/var/lib/rabbitmq
    restart: on-failure

  # Our custom django application. It includes Geonode.
  django:
    <<: *default-common-django
    entrypoint: [ "/entrypoint.sh" ]
    command: 'uwsgi --ini /uwsgi.conf'
#    volumes:
#      - ./volumes/statics:/home/web/statics:rw
    ports:
      - "3000:8080"

  nginx:
    image: nginx:1.17-alpine
    hostname: nginx
    volumes:
      - ./nginx/sites-enabled:/etc/nginx/conf.d:ro
      - ./nginx/pages:/pages:ro
      - ./volumes/logs:/var/log/nginx
      - ./volumes/statics/static:/home/web/static:ro
      - ./volumes/statics/media:/home/web/media:ro
      - ./volumes/statics/temp:/home/web/temp:ro
      - ./volumes/statics/gwml2:/home/web/gwml2:ro
    ports:
      - "${HTTP_PORT}:80"
    restart: unless-stopped

  # Geoserver backend
  geoserver:
    image: kartoza/geoserver:2.23.0-geo-v2023.09.15
    healthcheck:
      test: "curl --fail --silent --write-out 'HTTP CODE : %{http_code}\n' --output /dev/null http://127.0.0.1:8080/geoserver/rest/workspaces/geonode.html"
      interval: 60s
      timeout: 10s
      retries: 1
      start_period: 60s
    env_file:
      - .env
    volumes:
      - geodatadir:/spcgeonode-geodatadir/
      - ./geoserver/download.properties:/spcgeonode-geodatadir/download.properties
      - ./geoserver/entrypoint.sh:/usr/local/tomcat/tmp/entrypoint.sh
    ports:
      - "8080:8080"
    depends_on:
      - db
    restart: on-failure

  #    TODO:
  #     Comment this if using netcfd
  data-dir-conf:
    image: geonode/geoserver_data:2.23.0
    entrypoint: sleep infinity
    volumes:
      - geodatadir:/geoserver_data/data
    restart: on-failure
    healthcheck:
      test: "ls -A '/geoserver_data/data' | wc -l"

  # PostGIS database.
  db:
    # use geonode official postgis 10 image
    image: kartoza/postgis:16
    environment:
      # database info, ${variables} are read from .env
      - POSTGRES_DBNAME=${GEONODE_DATABASE},${GEONODE_GEODATABASE},groundwater
      - POSTGRES_USER=${GEONODE_DATABASE_USER}
      - POSTGRES_PASSWORD=${GEONODE_DATABASE_PASSWORD}
    volumes:
      - ./volumes/database:/var/lib/postgresql/16/
      - ./backups:/backups
    restart: on-failure
    ports:
      - "25432:5432"
    healthcheck:
      test: "pg_isready -d postgres -U postgres"

  # Celery worker that executes celery tasks created by Django.
  celery:
    <<: *default-common-django
    entrypoint: [ ]
    command: 'celery -A geonode.celery_app:app worker -B -E --statedb=worker.state -s celerybeat-schedule --loglevel=DEBUG --concurrency=10 -n worker1@%h'
    deploy:
      resources:
        limits:
          memory: '8G'
    depends_on:
      - rabbitmq

  # django application for development
  dev:
    <<: *default-common-django
    image: kartoza/igrac_ggis:${VERSION}-dev
    entrypoint: [ ]
    build:
      context: ../
      dockerfile: deployment/docker/Dockerfile
      target: dev
    depends_on:
      - geoserver
      - db
      - celery
    ports:
      # for django test server
      - "3002:8080"
      # for ssh
      - "3003:22"

  istsos:
    image: kartoza/igrac_ggis:istsos-${ISTSOS_VERSION}
    build:
      context: ../
      dockerfile: deployment/istsos/Dockerfile
    restart: always
    volumes:
      - ./istsos/services:/usr/share/istsos/services
      - ../istsos2/istsoslib:/usr/share/istsos/istsoslib:ro
      - ../istsos2/application_istsoslib.py:/usr/share/istsos/application_istsoslib.py:ro
    environment:
      - GROUNDWATER_DATABASE=groundwater
      - GROUNDWATER_DATABASE_USER=${GEONODE_DATABASE_USER}
      - GROUNDWATER_DATABASE_PASSWORD=${GEONODE_DATABASE_PASSWORD}
      - GROUNDWATER_DATABASE_HOST=${DATABASE_HOST}
      - GROUNDWATER_DATABASE_PORT=${DATABASE_PORT}
    ports:
      - "5001:80"

  minio:
    image: docker.io/bitnami/minio:latest
    ports:
      - '9000:9000'
      - '9001:9001'
    volumes:
      - ./volumes/minio:/bitnami/minio/data
    environment:
      - MINIO_ROOT_USER=minio-user
      - MINIO_ROOT_PASSWORD=minio-password
      - MINIO_DEFAULT_BUCKETS=ggis

volumes:
  database:
  geodatadir:
  certificates:
  pgdumps:
  rabbitmq: